[Serve] Log Serve config when controller receives it

shrekris-anyscale opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

<!-- Please give a short summary of the change and the problem this solves. -->

With this change, the controller logs the Serve config when it receives it.

## Related issue number

<!-- For example: "Closes #1234" -->

N/A

## Checks

- [X] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - This change relies on existing tests.

GeneDer reviewed (approved):
:shipit:

zcin reviewed (commented):
Could you include an example of what this would look like?

shrekris-anyscale commented:
> Could you include an example of what this would look like? Would this show up in the outputs streamed in `serve run`?

Good question– yes it show up in `serve run` if you pass in a config file. The output looks like this:

```
INFO 2024-01-26 12:31:39,553 controller 91367 controller.py:785 - Received config: proxy_location=<ProxyLocation.EveryNode: 'EveryNode'> http_options=HTTPOptionsSchema(host='0.0.0.0', port=8000, root_path='', request_timeout_s=None, keep_alive_timeout_s=5) grpc_options=gRPCOptionsSchema(port=9000, grpc_servicer_functions=[]) logging_config=None applications=[ServeApplicationSchema(name='app1', route_prefix='/', import_path='test_app:app', runtime_env={}, host='0.0.0.0', port=8000, deployments=[DeploymentSchema(name='hello', num_replicas=<DEFAULT.VALUE: 1>, route_prefix=<DEFAULT.VALUE: 1>, max_concurrent_queries=<DEFAULT.VALUE: 1>, user_config=<DEFAULT.VALUE: 1>, autoscaling_config=<DEFAULT.VALUE: 1>, graceful_shutdown_wait_loop_s=<DEFAULT.VALUE: 1>, graceful_shutdown_timeout_s=<DEFAULT.VALUE: 1>, health_check_period_s=<DEFAULT.VALUE: 1>, health_check_timeout_s=<DEFAULT.VALUE: 1>, ray_actor_options=<DEFAULT.VALUE: 1>, placement_group_bundles=<DEFAULT.VALUE: 1>, placement_group_strategy=<DEFAULT.VALUE: 1>, max_replicas_per_node=<DEFAULT.VALUE: 1>, logging_config=<DEFAULT.VALUE: 1>)], args={}, logging_config=None)] target_capacity=None
```

I'll look into ways to clean this up.

shrekris-anyscale commented:
I added `extra={"log_to_stderr": False}` to the log statement, so it doesn't show up in `serve run`. I also modified the formatting, so the config looks like this:

```
INFO 2024-01-26 12:55:26,239 controller 97705 controller.py:786 - Received config:
{
    "proxy_location": "EveryNode",
    "http_options": {
        "host": "0.0.0.0",
        "port": 8000
    },
    "grpc_options": {
        "port": 9000,
        "grpc_servicer_functions": []
    },
    "applications": [
        {
            "name": "app1",
            "route_prefix": "/",
            "import_path": "test_app:app",
            "runtime_env": {},
            "deployments": [
                {
                    "name": "hello"
                }
            ]
        }
    ]
}
```

edoakes commented:
This is going to be extremely noisy and will not format well if you use multi-line formatting. Is there a way we can distill the information we care about?

shrekris-anyscale commented:
> This is going to be extremely noisy

I updated the PR to only write a log when a new config is received.

shrekris-anyscale commented:
This does look pretty confusing with JSON mode enabled though. There are lots of newlines and spaces:

```
{"levelname": "INFO", "asctime": "2024-01-29 11:19:58,221", "component_name": "controller", "component_id": "98998", "message": "controller.py:791 - Deploying a new config:\n{\n    \"proxy_location\": \"EveryNode\",\n    \"http_options\": {\n        \"host\": \"0.0.0.0\",\n        \"port\": 8000\n    },\n    \"grpc_options\": {\n        \"port\": 9000,\n        \"grpc_servicer_functions\": []\n    },\n    \"logging_config\": {\n        \"encoding\": \"JSON\",\n        \"log_level\": \"INFO\",\n        \"logs_dir\": null,\n        \"enable_access_log\": true\n    },\n    \"applications\": [\n        {\n            \"name\": \"app1\",\n            \"route_prefix\": \"/\",\n            \"import_path\": \"test_app:app\",\n            \"runtime_env\": {},\n            \"deployments\": [\n                {\n                    \"name\": \"hello\"\n                }\n            ]\n        }\n    ]\n}"}
```

shrekris-anyscale commented:
I took a deeper look at the existing logs. We already log a message when the code version of an app has changed ([code link](https://github.com/ray-project/ray/blob/d2311655cf43df629c2e7123414b9e1fa77707a9/python/ray/serve/_private/application_state.py#L414)). This should be sufficient to detect when an entirely new config has been submitted.

Given this, there isn't a pressing reason to log the entire config. I'll close this PR for now.
